### PR TITLE
Changed the Type_Index to the new ComponentType

### PR DESCRIPTION
**Pull request by TheRealChriAnd**
* ECS Components does not need a .cpp file anymore

### CODE REVIEW

File: LambdaEngine/Include/ECS/ComponentType.h [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+		const ConstString m_ClassName;
+	};
+
+	struct ComponentTypeHasher
+	{
+		size_t operator()(const ComponentType* key) const
+		{
+			return key->GetHash();
+		}
+	};
```
Comment (Mumsfilibaba) [inline LambdaEngine/Include/ECS/ComponentType.h]:
```suggestion
```

**Review by Mumsfilibaba — APPROVED** [review]
👍

File: LambdaEngine/Include/ECS/ComponentType.h (hunk 1 above)
Comment (TheRealChriAnd) [reply LambdaEngine/Include/ECS/ComponentType.h]: Nice catch :)

Comment (TheRealChriAnd) [reply LambdaEngine/Include/ECS/ComponentType.h]: Nice catch :)

File: LambdaEngine/Include/ECS/ECSCore.h [hunk 2]
```diff
@@ -112,9 +113,9 @@ namespace LambdaEngine
 		/*	Create component immediately, but hold off on registering and publishing it until the end of the frame.
 			This is to prevent concurrency issues. Publishing a component means pushing entity IDs to IDVectors,
 			and there is no guarentee that no one is simultaneously reading from these IDVectors. */
-		m_ComponentsToRegister.PushBack({entity, Comp::s_TID});
-
-		return m_ComponentStorage.AddComponent<Comp>(entity, component);
+		Comp& comp = m_ComponentStorage.AddComponent<Comp>(entity, component);
```
Comment (TheoBerlin) [inline LambdaEngine/Include/ECS/ECSCore.h]: Is there a reason you changed the order of adding the component and enqueuing its registration?

File: LambdaEngine/Include/ECS/ComponentStorage.h [hunk 3]
```diff
@@ -25,32 +25,32 @@ namespace LambdaEngine
...
-		const IComponentArray* GetComponentArray(std::type_index componentType) const;
+		IComponentArray* GetComponentArray(const ComponentType* pComponentType);
+		const IComponentArray* GetComponentArray(const ComponentType* pComponentType) const;
 
 		template<typename Comp>
 		ComponentArray<Comp>* GetComponentArray();
 
 	private:
-		std::unordered_map<std::type_index, uint32> m_CompTypeToArrayMap;
+		std::unordered_map<const ComponentType*, uint32> m_CompTypeToArrayMap;
```
Comment (TheoBerlin) [inline LambdaEngine/Include/ECS/ComponentStorage.h:42]: Is `ComponentType::operator()` or the pointer being used for mapping?

File: LambdaEngine/Include/ECS/ComponentType.h [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+}
+
+namespace std
+{
+	template<>
+	struct hash<const LambdaEngine::ComponentType*>
+	{
+		size_t operator()(const LambdaEngine::ComponentType* componentType) const
+		{
+			return componentType->GetHash();
```
Comment (TheoBerlin) [inline LambdaEngine/Include/ECS/ComponentType.h]:
```suggestion
			return pComponentType->GetHash();
```

File: LambdaEngine/Include/ECS/ECSCore.h [hunk 5]
```diff
@@ -82,7 +83,7 @@ namespace LambdaEngine
 		void PerformComponentRegistrations();
 		void PerformComponentDeletions();
 		void PerformEntityDeletions();
-		bool DeleteComponent(Entity entity, std::type_index componentType);
+		bool DeleteComponent(Entity entity, const ComponentType* componentType);
```
Comment (TheoBerlin) [inline LambdaEngine/Include/ECS/ECSCore.h]:
```suggestion
		bool DeleteComponent(Entity entity, const ComponentType* pComponentType);
```

File: LambdaEngine/Include/ECS/ECSCore.h (hunk 2 above)
Comment (TheRealChriAnd) [reply LambdaEngine/Include/ECS/ECSCore.h]: I have not actually changed that. Maybe when I merged from master? Or the change was made on the master while I was working on this branch? Either way, let me know what is the correct code.

File: LambdaEngine/Source/ECS/ECSCore.cpp [hunk 6]
```diff
@@ -143,7 +144,7 @@ namespace LambdaEngine
 		m_EntitiesToDelete.Clear();
 	}
 
-	bool ECSCore::DeleteComponent(Entity entity, std::type_index componentType)
+	bool ECSCore::DeleteComponent(Entity entity, const ComponentType* componentType)
```
Comment (TheoBerlin) [inline LambdaEngine/Source/ECS/ECSCore.cpp]:
```suggestion
	bool ECSCore::DeleteComponent(Entity entity, const ComponentType* pComponentType)
```

File: LambdaEngine/Include/ECS/ComponentStorage.h (hunk 3 above)
Comment (TheRealChriAnd) [reply LambdaEngine/Include/ECS/ComponentStorage.h:42]: Yes the `ComponentType::operator()` is the hasher used by default

**Review by TheoBerlin — CHANGES_REQUESTED** [review]
Found some missing `p` prefixes.

File: LambdaEngine/Include/ECS/ECSCore.h (hunk 2 above)
Comment (TheoBerlin) [reply LambdaEngine/Include/ECS/ECSCore.h]: Do it the 'old' way, register.PushBack first, then return storage.AddComp

**Review by TheoBerlin — APPROVED** [review]

**Pull request merged**